Add try-catch to finding first agent instance in multi_agent_types!

tlow22 opened the pull request:
When initializing the `adata` dataframe for multi-agents, code tries to look for the first object instance of the different agent types., so that it can be used to extract the desired property types for dataframe header. 

In the specific case where one agent type doesn't yet exist at model initialization (to be added later on while stepping the simulation), this will cause an error. 

The behavior now is to default the missing agent type's properties to `Missing`. If desired, I can also leverage `hasfield` and `fieldtype` when the desired property is a `Symbol`

Datseris commented:
How does `missing` come into play, given that the code returns `nothing` if the `try` fails?

tlow22 commented:
@Datseris, ah I'm leveraging some of the native code a few more lines below the change, namely lines 250 - 259: 

``` 
if k isa Symbol
    current_type =
        hasproperty(a, k) ? typeof(get_data(a, k, identity)) : Missing
else
    current_type = try
        typeof(get_data(a, k, identity))
    catch
        Missing
    end
end
```

Even if an agent instance is not found, and `nothing` is returned, the code continues to get the object properties either via the `Symbol` or `function` getter method. I noticed that in both cases, a `try-catch` with `Missing` is being returned when the property getter was used. If the object is `nothing`, the property getters will fail and simply default to `Missing`.

Datseris commented:
Have you measured what impact on performance this may have? A typical `run!` call of one of the working examples of the repo. How much slower does it become on this branch?

Datseris commented:
Also, what happens when you actually use this in practice? The data frame is initialized with a Missing column, and once you try to add data once agents exist, what happens? 

This Pull Request does not have tests. Please add a test that test this functionalityt by making a minimal ABM that doesn't have agents and then goes through `run!` so that we have proof that the code works as exepected. The test must be added in the `test` subfolder. Have a look at the files already there and you'll get the idea of how it works.

tlow22 commented:
@Datseris Took your comment into consideration and added functionality to determine the `fieldtype` when an object is not available. Added unit test to verify 

1) fieldtype determination 
2) successful addition to the dataframe when agents are introduced later on.  

Caveat - same trick can't be applied to functions or aggregators, as there are too many ways a function could fail. 

PR shouldn't affect performance, as it only affects the dataframe initialization right at the very beginning and not subsequent runs.

Datseris commented:
Thanks a lot @tlow22 . Can you please add a note in the changelog under the nextx minor version about this fix?

---

> PR shouldn't affect performance, as it only affects the dataframe initialization right at the very beginning and not subsequent runs

Sure, sounds reasonable, but can you please check explicitly?

tlow22 commented:
@Datseris sure thing on changelog update! Here's a timing benchmark comparison for the PR: 

Main: 
![image](https://user-images.githubusercontent.com/32314421/209567378-597c0164-8523-4bcb-a751-3711fac93829.png)

PR: 
![image](https://user-images.githubusercontent.com/32314421/209567403-b033256f-1f0a-4989-8685-b03b273c24cc.png)

Datseris reviewed (approved):
Thank you very much for the contribution @tlow22  :)

tlow22 commented:
@Datseris glad to help in a little way! Recently discovered DrWatson as well, very cool work!
